File: src/python/cable_joints/pbd_resolve_cable_over_corrections.py
import numpy as np
from collections import defaultdict

from .ecs import World
from .cable_joints_components import (
    CableJointComponent,
    CablePathComponent,
)
from .ecs import (
    PositionComponent,
    MassComponent,
    MomentOfInertiaComponent,
    OrientationComponent,
)
from .update_attachment_points import calculate_attachment_points


class PBDResolveCableOverCorrections:
    """
    This system runs after the main PBD cable solver to correct for
    over-contraction artifacts. When multiple taut cables pull on the same
    object, the main solver can cause joints that should remain taut to become
    slack. This pass identifies such joints and pushes them back towards
    their rest length.

    It uses a Jacobi-style approach, calculating all corrections based on the
    state after the main solver, and then applying an averaged correction for
    each body. This prevents the instability of a sequential solve and better
    handles bodies affected by multiple over-corrected joints.
    """
    run_in_pause = False

    def update(self, world: World, _dt_unused):
        path_entities = world.query(CablePathComponent)
        if not path_entities:
            return

        joint_to_path_and_index = {}
        all_joint_ids = set()
        for path_id in path_entities:
            path = world.get_component(path_id, CablePathComponent)
            for i, joint_id in enumerate(path.joint_entities):
                all_joint_ids.add(joint_id)
                joint_to_path_and_index[joint_id] = (path, i)

        # 1. Find all joints that were taut but are now slack
        over_corrected_joints = []
        for joint_id in all_joint_ids:
            joint = world.get_component(joint_id, CableJointComponent)
            p_a = joint.attachment_point_a_world
            p_b = joint.attachment_point_b_world
            pre_cable_solve_length = np.linalg.norm(p_a - p_b)
            if pre_cable_solve_length >= joint.rest_length: # If was stretched
                path, i = joint_to_path_and_index[joint_id]
                p_a, p_b = calculate_attachment_points(world, joint, path, i)
                post_cable_solve_length = np.linalg.norm(p_a - p_b)
                if post_cable_solve_length < joint.rest_length: # If now slack
                    over_corrected_joints.append(joint_id)

        if not over_corrected_joints:
            return

        # Jacobi-style solve: calculate all corrections first, then apply averaged corrections.
        position_corrections = defaultdict(list)
        angle_corrections = defaultdict(list)

        # 2. Calculate corrections for each over-corrected joint without applying them
        for joint_id in over_corrected_joints:
            self.calculate_joint_correction(
                world, joint_id, joint_to_path_and_index, position_corrections, angle_corrections
            )

        # 3. Apply the averaged corrections to each affected entity
        for entity_id, pos_deltas in position_corrections.items():
            if pos_deltas:
                avg_delta = np.mean(pos_deltas, axis=0)
                pos_comp = world.get_component(entity_id, PositionComponent)
                if pos_comp:
                    pos_comp.pos += avg_delta

        for entity_id, ang_deltas in angle_corrections.items():
            if ang_deltas:
                avg_delta = np.mean(ang_deltas)
                orientation_comp = world.get_component(entity_id, OrientationComponent)
                if orientation_comp:
                    orientation_comp.angle += avg_delta

    def calculate_joint_correction(self, world, joint_id, joint_to_path_and_index, position_corrections, angle_corrections):
        """
        Calculates the PBD correction for a single joint that has become slack
        and stores the delta values in the provided dictionaries instead of
        applying them directly to the components.
        """
        joint = world.get_component(joint_id, CableJointComponent)
        path_data = joint_to_path_and_index.get(joint_id)
        if not path_data:
            return
        path, i = path_data

        p_a, p_b = calculate_attachment_points(world, joint, path, i)
        if p_a is None or p_b is None:
            return

        current_segment_length = np.linalg.norm(p_a - p_b)
        constraint_error = current_segment_length - joint.rest_length

        epsilon = 1e-9
        # This function should only be called for slack joints, where error is negative.
        if constraint_error >= -epsilon:
            return

        entity_a = joint.entity_a
        entity_b = joint.entity_b

        mass_a_comp = world.get_component(entity_a, MassComponent)
        inv_mass_a = 1.0 / mass_a_comp.mass if mass_a_comp and mass_a_comp.mass > 0 and np.isfinite(mass_a_comp.mass) else 0.0
        moi_a_comp = world.get_component(entity_a, MomentOfInertiaComponent)
        inv_inertia_a = moi_a_comp.inv_inertia if moi_a_comp else 0.0

        mass_b_comp = world.get_component(entity_b, MassComponent)
        inv_mass_b = 1.0 / mass_b_comp.mass if mass_b_comp and mass_b_comp.mass > 0 and np.isfinite(mass_b_comp.mass) else 0.0
        moi_b_comp = world.get_component(entity_b, MomentOfInertiaComponent)
        inv_inertia_b = moi_b_comp.inv_inertia if moi_b_comp else 0.0

        if inv_mass_a + inv_mass_b + inv_inertia_a + inv_inertia_b <= epsilon:
            return

        diff = p_b - p_a
        length = np.linalg.norm(diff)
        if length <= epsilon:
            return
        direction = diff / length

        grad_pos_a = -direction
        grad_pos_b = direction

        pos_a_comp = world.get_component(entity_a, PositionComponent)
        r_a = p_a - pos_a_comp.pos
        r_a_3d = np.array([r_a[0], r_a[1], 0.0])
        direction_3d = np.array([direction[0], direction[1], 0.0])
        grad_ang_a = np.cross(r_a_3d, direction_3d)[2]

        pos_b_comp = world.get_component(entity_b, PositionComponent)
        r_b = p_b - pos_b_comp.pos
        r_b_3d = np.array([r_b[0], r_b[1], 0.0])
        neg_direction_3d = np.array([-direction[0], -direction[1], 0.0])
        grad_ang_b = np.cross(r_b_3d, neg_direction_3d)[2]

        denom = 0.0
        denom += inv_mass_a * np.dot(grad_pos_a, grad_pos_a)
        denom += inv_inertia_a * grad_ang_a * grad_ang_a
        denom += inv_mass_b * np.dot(grad_pos_b, grad_pos_b)
        denom += inv_inertia_b * grad_ang_b * grad_ang_b

        dt = world.get_resource('dt')
        if dt is not None and dt > 0:
            denom += path.compliance / (dt * dt)

        if denom <= epsilon:
            return

        lambda_ = -constraint_error / denom

        # Store corrections instead of applying them directly
        if inv_mass_a > 0.0:
            delta_pos_a = grad_pos_a * (inv_mass_a * lambda_)
            position_corrections[entity_a].append(delta_pos_a)

        if inv_inertia_a > 0.0:
            delta_ang_a = -inv_inertia_a * lambda_ * grad_ang_a
            angle_corrections[entity_a].append(delta_ang_a)

        if inv_mass_b > 0.0:
            delta_pos_b = grad_pos_b * (inv_mass_b * lambda_)
            position_corrections[entity_b].append(delta_pos_b)

        if inv_inertia_b > 0.0:
            delta_ang_b = -inv_inertia_b * lambda_ * grad_ang_b
            angle_corrections[entity_b].append(delta_ang_b)

File: src/js/cable_joints/pbdResolveCableOverCorrections.js
import Vector2 from './vector2.js';
import {
  CableJointComponent,
  CablePathComponent,
  calculateAttachmentPoints
} from './cable_joints_core.js';
import {
  PositionComponent,
  MassComponent,
  MomentOfInertiaComponent,
  OrientationComponent
} from './ecs.js';

function ensureArrayMapEntry(map, key, value) {
  if (!map.has(key)) {
    map.set(key, []);
  }
  if (value !== undefined) {
    map.get(key).push(value);
  }
}

export class PBDResolveCableOverCorrections {
  runInPause = false;

  update(world, _dt_unused) {
    const pathEntities = world.query([CablePathComponent]);
    if (!pathEntities) return;
    const diag = {
      overCorrectedJointCount: 0,
      entitiesPosAdjusted: 0,
      entitiesAngAdjusted: 0,
      maxPosAdjustment: 0.0,
      maxAngAdjustment: 0.0
    };

    const jointToPathAndIndex = new Map();
    const allJointIds = new Set();

    for (const pathId of pathEntities) {
      const path = world.getComponent(pathId, CablePathComponent);
      if (!path.jointEntities || path.jointEntities.length === 0) continue;

      for (let i = 0; i < path.jointEntities.length; i++) {
        const jointId = path.jointEntities[i];
        allJointIds.add(jointId);
        jointToPathAndIndex.set(jointId, { path, i });
      }
    }

    const overCorrected = [];
    for (const jointId of allJointIds) {
      const joint = world.getComponent(jointId, CableJointComponent);
      const preLen = joint.attachmentPointA_world.distanceTo(joint.attachmentPointB_world);
      if (preLen >= joint.restLength) {
        const data = jointToPathAndIndex.get(jointId);
        if (!data) continue;
        const { path, i } = data;
        const { attachmentA_current: pA, attachmentB_current: pB } = calculateAttachmentPoints(world, joint, path, i);
        if (!pA || !pB) continue;
        const postLen = pA.distanceTo(pB);
        if (postLen < joint.restLength) {
          overCorrected.push(jointId);
        }
      }
    }

    diag.overCorrectedJointCount = overCorrected.length;
    if (overCorrected.length < 2) {
      world.setResource('cableOverCorrectionDiag', diag);
      return;
    }

    const posCorrections = new Map();
    const angCorrections = new Map();

    for (const jointId of overCorrected) {
      this.calculateJointCorrection(world, jointId, jointToPathAndIndex, posCorrections, angCorrections);
    }

    for (const [entityId, deltas] of posCorrections.entries()) {
      if (deltas.length >= 2) {
        // Multiple joints connected to this one body are slack now
        // They should not have been considered to have such a great effect on
        // the body during this time step. Correct that back by averaging.
        const avg = deltas.reduce((acc, v) => acc.add(v), new Vector2()).scale(1 / deltas.length);
        const posComp = world.getComponent(entityId, PositionComponent);
        if (posComp) {
          posComp.pos.add(avg);
          diag.entitiesPosAdjusted += 1;
          diag.maxPosAdjustment = Math.max(diag.maxPosAdjustment, avg.length());
        }
      }
    }

    for (const [entityId, deltas] of angCorrections.entries()) {
      if (deltas.length >= 2) {
        // Multiple joints connected to this one body are slack now
        // They should not have been considered to have such a great effect on
        // the body during this time step. Correct that back by averaging.
        const sum = deltas.reduce((a, b) => a + b, 0);
        const avg = sum / deltas.length;
        const orientComp = world.getComponent(entityId, OrientationComponent);
        if (orientComp) {
          orientComp.angle += avg;
          diag.entitiesAngAdjusted += 1;
          diag.maxAngAdjustment = Math.max(diag.maxAngAdjustment, Math.abs(avg));
        }
      }
    }
    world.setResource('cableOverCorrectionDiag', diag);
  }

  calculateJointCorrection(world, jointId, jointToPathAndIndex, posCorrections, angCorrections) {
    const joint = world.getComponent(jointId, CableJointComponent);
    const data = jointToPathAndIndex.get(jointId);
    if (!data) return;
    const { path, i } = data;

    const { attachmentA_current: pA, attachmentB_current: pB } = calculateAttachmentPoints(world, joint, path, i);
    if (!pA || !pB) return;

    const currentLen = pA.distanceTo(pB);
    const constraintError = currentLen - joint.restLength;
    const epsilon = 1e-9;
    if (constraintError >= -epsilon) return;

    const entityA = joint.entityA;
    const entityB = joint.entityB;

    const massAComp = world.getComponent(entityA, MassComponent);
    const invMassA = massAComp && massAComp.mass > 0 && Number.isFinite(massAComp.mass) ? 1.0 / massAComp.mass : 0.0;
    const moiAComp = world.getComponent(entityA, MomentOfInertiaComponent);
    const invInertiaA = moiAComp ? moiAComp.invInertia : 0.0;

    const massBComp = world.getComponent(entityB, MassComponent);
    const invMassB = massBComp && massBComp.mass > 0 && Number.isFinite(massBComp.mass) ? 1.0 / massBComp.mass : 0.0;
    const moiBComp = world.getComponent(entityB, MomentOfInertiaComponent);
    const invInertiaB = moiBComp ? moiBComp.invInertia : 0.0;

    if (invMassA + invMassB + invInertiaA + invInertiaB <= epsilon) return;

    const diff = new Vector2().subtractVectors(pB, pA);
    const len = diff.length();
    if (len <= epsilon) return;
    const dir = diff.clone().scale(1 / len);

    const gradPosA = dir.clone().scale(-1);
    const gradPosB = dir.clone();

    const posAComp = world.getComponent(entityA, PositionComponent);
    const rA = new Vector2().subtractVectors(pA, posAComp.pos);
    const gradAngA = rA.x * dir.y - rA.y * dir.x;

    const posBComp = world.getComponent(entityB, PositionComponent);
    const rB = new Vector2().subtractVectors(pB, posBComp.pos);
    const gradAngB = rB.x * (-dir.y) - rB.y * (-dir.x);

    let denom = 0;
    denom += invMassA * gradPosA.lengthSq();
    denom += invInertiaA * gradAngA * gradAngA;
    denom += invMassB * gradPosB.lengthSq();
    denom += invInertiaB * gradAngB * gradAngB;

    const dt = world.getResource('dt');
    if (dt !== undefined && dt > 0) {
      denom += path.compliance / (dt * dt);
    }

    if (denom <= epsilon) return;

    const lambda = -constraintError / denom;

    if (invMassA > 0) {
      const delta = gradPosA.clone().scale(invMassA * lambda);
      ensureArrayMapEntry(posCorrections, entityA, delta);
    }
    if (invInertiaA > 0) {
      const delta = -invInertiaA * lambda * gradAngA;
      ensureArrayMapEntry(angCorrections, entityA, delta);
    }
    if (invMassB > 0) {
      const delta = gradPosB.clone().scale(invMassB * lambda);
      ensureArrayMapEntry(posCorrections, entityB, delta);
    }
    if (invInertiaB > 0) {
      const delta = -invInertiaB * lambda * gradAngB;
      ensureArrayMapEntry(angCorrections, entityB, delta);
    }
  }
}
